text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+
+```rust
+func(x is Some(y) && y > 3);
+```
+
+The `is` operator may not appear as a statement; use `let` to bind a pattern in
+that context:
+
+```rust
+// ERROR: use `let` for this
+an_expression() is x;
+
+let x = an_expression();
+```
+
+# Reference-level explanation
+
+Add a new [operator
+expression](https://doc.rust-lang.org/reference/expressions/operator-expr.html),
+`IsExpression`:
+
+> **<sup>Syntax</sup>**\
+> _IsExpression_ :\
+> &nbsp;&nbsp; _Expression_ `is` _PatternNoTopAlt_
+
+Add `is` to the [operator
+precedence](https://doc.rust-lang.org/reference/expressions.html#expression-precedence)
+table, at the same precedence level as `==`, and likewise non-associative
+(requiring parentheses).
+
+Detect `is` appearing as a top-level statement and produce an error, with a
+rustfix suggestion to use `let` instead.
+
+# Drawbacks
+
+Introducing both the `is` operator and `let`-chains would provide two different
+ways to do a pattern match as part of a condition. Having more than one way to
+do something could lead people to wonder if there's a difference; we would need
+to clearly communicate that they serve similar purposes.
+
+An `is` operator will produce a name conflict with [the `is` method on
+`dyn Any`](https://doc.rust-lang.org/std/any/trait.Any.html#method.is) in the
+standard library, and with the (relatively few) methods named `is` in the
+ecosystem. This will not break any existing Rust code, as the operator will
+only exist in the Rust 2024 edition and newer. The Rust standard library and
+any other library that wants to avoid requiring the use of `r#is` in Rust 2024
+and newer could provide aliases of these methods under a new name; for
+instance, the standard library could additionally provide `Any::is` under a new
+name `is_type`.
+
+# Rationale and alternatives
+
+As noted in the [motivation](#motivation) section, adding the `is` operator
+allows writing pattern matches from left-to-right, which reads more naturally
+in some conditionals and fits well with method chains and similar. As noted
+under [prior art](#prior-art), other languages such as C# already have this
+exact operator for this exact purpose.
+
+We could choose not to add this operator, and have *only* `let`-chains. This
+would provide equivalent functionality, semantically; however, it would force
+pattern-matches to be written with the pattern on the left, which won't read as
+naturally in some expressions. Notably, this seems unlikely to do as effective
+a job of reducing the desire for `is_variant()` methods and helpers like
+`is_some_and(...)`.
+
+We could choose to add the `is` operator and *not* add `let`-chains. However,
+many people *already* expect `let`-chains to work as an obvious extrapolation
+from seeing `if let`/`while let` syntax.
+
+We could add this operator using punctuation instead (e.g. `~`). However, there
+is no "natural" operator that conveys "pattern match" to people (the way that
```

Comment:
I disagree that there's no natural operator, even though I agree that it would be less clear.

For example, we could use tildes as an additional equality operator (`x ~ Some(y)`) and it would fit in with Rust fine. I mostly say this because I think that the reasoning should lean heavier on its stronger arguments and not for a lack of creativity:

* `is` is short and immediately clear
* Any use of it as a variable is something that won't be missed (e.g. as a plural of `i`, an already nameless variable)
* People are already used to it being a keyword in other languages, so, it being one in Rust too isn't strange

Reply:
Something else to point to, `is` is easy to write too, which is worth considering. 
For example it annoys me to have to write `#[(...)]`
In my opinion we should try to avoid symbols where words suffice.

Reply:
For some more prior art, Raku uses [`~~`](https://docs.raku.org/routine/~~) for “smart matching”. Using two tildes would leave the single tilde free for something else, if we wanted to use it later. Of course, Raku is known to be a symbol-heavy language, so I don’t think this is the best choice.

Reply:
I'd like to note that Rust _used_  to use tildes (`~T`), and during the move to `Box<T>` it was noted that  `~` is simply _absent_ (not just hard to type, plain absent) from a number of keyboard layouts.

As an example, consider a [Polish keyboard layout](https://commons.wikimedia.org/wiki/File:KB_Polish_QWERTZ_PN87_commons.svg).

I would recommend avoiding `~` altogether.

Reply:
@matthieu-m The keyboard layout you've linked to is an obsolete typewriter layout. Polish computers use a QWERTY-based layout called ["Polish programmer's" layout](https://commons.wikimedia.org/wiki/File:Polish_programmer%27s_layout.PNG), which despite the name, is the default used by everyone.

Reply:
another argument that i have yet to see is that the `is` keyword fits very well with the `as` keyword already present in Rust